internal/listtoken/list_token.go

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+// Package listtoken encapsulates domain logic surrounding
+// list endpoint refresh tokens. Refresh tokens are used when users
+// paginate through results in our list endpoints, and also to
+// allow users to request new, updated and deleted resources.
+package listtoken
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/db/timestamp"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/types/resource"
+)
+
+// A Token is returned in list endpoints for the purposes of pagination
+type Token struct {
+	CreateTime   time.Time
+	ResourceType resource.Type
+	GrantsHash   []byte
+	Subtype      TokenSubtype
+}
+
+type TokenSubtype interface {
+	isTokenSubtype()
+}
+
+type PaginationToken struct {
+	LastItemId         string
+	LastItemCreateTime time.Time
+}
+
+func (*PaginationToken) isTokenSubtype() {}
+
+type StartRefreshToken struct {
+	PhaseEndTime time.Time
+	LastUseTime  time.Time
+}
+
+func (*StartRefreshToken) isTokenSubtype() {}
+
+type RefreshToken struct {
+	PhaseStartTime     time.Time
+	PhaseEndTime       time.Time
+	LastUseTime        time.Time
+	LastItemId         string
+	LastItemUpdateTime time.Time
+}
```

Comment:
The StartRefreshToken and RefreshToken types are very similar - indeed, StartRefreshToken is just a subset of the RefreshToken. We made these two different types because the alternative would have been to add many optional fields to the RefreshToken. This way, we can use strict validation of the types at construction time and carry those guarantees through the domain logic.